pkg/app/ops/deploymentchaincontroller/controller.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2021 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deploymentchaincontroller
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/datastore"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+const (
+	interval = 1 * time.Minute
+)
+
+type DeploymentChainController struct {
+	applicationStore     datastore.ApplicationStore
+	deploymentStore      datastore.DeploymentStore
+	deploymentChainStore datastore.DeploymentChainStore
+	// Map from deployment chain ID to the updater
+	// who in charge for the deployment chain updating.
+	updaters map[string]*updater
+	// WaitGroup for waiting all deployment chain updaters to be completed.
+	wg sync.WaitGroup
+
+	logger *zap.Logger
+}
+
+func NewDeploymentChainController(
+	ds datastore.DataStore,
+	logger *zap.Logger,
+) *DeploymentChainController {
+	return &DeploymentChainController{
+		applicationStore:     datastore.NewApplicationStore(ds),
+		deploymentStore:      datastore.NewDeploymentStore(ds),
+		deploymentChainStore: datastore.NewDeploymentChainStore(ds),
+		updaters:             make(map[string]*updater),
+		logger:               logger.Named("deployment-chain-controller"),
+	}
+}
+
+func (d *DeploymentChainController) Run(ctx context.Context) error {
+	d.logger.Info("start running DeploymentChainController")
+
+	t := time.NewTicker(interval)
+	defer t.Stop()
+	d.logger.Info("start syncing updaters")
+
+L:
+	for {
+		select {
+		case <-ctx.Done():
+			break L
+
+		case <-t.C:
+			d.syncUpdaters(ctx)
+		}
+	}
+
+	d.logger.Info("deploymentChainController has been stopped")
```

Comment:
nit
```suggestion
	defer d.logger.Info("deploymentChainController has been stopped")
	for {
		select {
		case <-ctx.Done():
			return nil

		case <-t.C:
			d.syncUpdaters(ctx)
		}
	}
```

Reply:
I would prefer to use Label as little as possible, but it's on you.

Reply:
Sure, I was overthinking there 👍